pkg/ingester/encoding.go

```diff
@@ -25,84 +25,145 @@ const (
 	buf := EncWith(b)
 	buf.PutByte(byte(WALRecordSeries))
-	buf.PutUvarintStr(record.UserID)
+	buf.PutUvarintStr(r.UserID)
 
-	var enc tsdb_record.Encoder
+	var enc record.Encoder
 	// The 'encoded' already has the type header and userID here, hence re-using
 	// the remaining part of the slice (i.e. encoded[len(encoded):])) to encode the series.
 	encoded := buf.Get()
-	encoded = append(encoded, enc.Series(record.Series, encoded[len(encoded):])...)
+	encoded = append(encoded, enc.Series(r.Series, encoded[len(encoded):])...)
 
 	return encoded
 }
 
-func (record *WALRecord) encodeEntries(b []byte) []byte {
+func (r *WALRecord) encodeEntries(b []byte) []byte {
 	buf := EncWith(b)
 	buf.PutByte(byte(WALRecordEntries))
-	buf.PutUvarintStr(record.UserID)
-
-	entries := record.RefEntries.Entries
-	if len(entries) == 0 {
-		return buf.Get()
+	buf.PutUvarintStr(r.UserID)
+
+	// Placeholder for the first timestamp of any sample encountered.
+	// All others in this record will store their timestamps as diffs relative to this
+	// as a space optimization.
+	var first int64
+
+outer:
+	for _, ref := range r.RefEntries {
+		for _, entry := range ref.Entries {
+			first = entry.Timestamp.UnixNano()
+			buf.PutBE64int64(first)
+			break outer
+		}
 	}
 
-	// Only encode the series fingerprint if there are >0 entries.
-	buf.PutBE64(record.RefEntries.Ref)
-
-	// Store base timestamp and base reference number of first sample.
-	// All samples encode their timestamp and ref as delta to those.
-	first := entries[0].Timestamp.UnixNano()
-
-	buf.PutBE64int64(first)
-
-	for _, s := range entries {
-		buf.PutVarint64(s.Timestamp.UnixNano() - first)
-		// denote line length
-		byteLine := []byte(s.Line)
-		buf.PutUvarint(len(byteLine))
-		buf.PutBytes(byteLine)
+	for _, ref := range r.RefEntries {
+		// ignore refs with 0 entries
+		if len(ref.Entries) < 1 {
+			continue
+		}
+		buf.PutBE64(ref.Ref)             // write fingerprint
+		buf.PutUvarint(len(ref.Entries)) // write number of entries
+
+		for _, s := range ref.Entries {
+			buf.PutVarint64(s.Timestamp.UnixNano() - first)
+			// denote line length
+			byteLine := []byte(s.Line)
+			buf.PutUvarint(len(byteLine))
+			buf.PutBytes(byteLine)
+		}
 	}
 	return buf.Get()
 
 }
 
-func decodeEntries(b []byte, entries *RefEntries) error {
+func decodeEntries(b []byte, rec *WALRecord) error {
 	if len(b) == 0 {
 		return nil
 	}
 
 	dec := DecWith(b)
-
-	entries.Ref = dec.Be64()
 	baseTime := dec.Be64int64()
 
 	for len(dec.B) > 0 && dec.Err() == nil {
-		dRef := dec.Varint64()
-		ln := dec.Uvarint()
-		line := dec.Bytes(ln)
-
-		entries.Entries = append(entries.Entries, logproto.Entry{
-			Timestamp: time.Unix(0, baseTime+dRef),
-			Line:      string(line),
-		})
+		refEntries := RefEntries{
+			Ref:     dec.Be64(),
+			Entries: recordPool.GetEntries(),
+		}
+
+		nEntries := dec.Uvarint()
+		rem := nEntries
+		if cap(refEntries.Entries) < nEntries {
+			refEntries.Entries = make([]logproto.Entry, 0, nEntries)
```

Comment:
I think you should not put back into the pool things that you didn't get from the pool. But I can't find what outcome it will have, it is just something no one does.

Basically I wouldn't make here, I would let the slice to grow and when you put back decide or not if you should put it back based on it;s size, which the leaky bucket implementation does for you.